compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+    body: &mut Body<'tcx>,
+    tcx: TyCtxt<'tcx>,
+    param_env: ParamEnv<'tcx>,
+) -> bool {
+    let mut visitor = AnalysisVisitor {
+        analysis: IndexVec::from_elem_n(Analysis::default(), body.local_decls.len()),
+    };
+    visitor.visit_body(body);
+    let analysis = visitor.analysis;
+
+    let mut invalidated_statements = FxHashSet::default();
+
+    'outer: for (_local, result) in analysis.iter_enumerated() {
+        let Some((temp_loc, (temp_place, temp_rvalue))) = as_assign(result.write, body) else {
+            continue;
+        };
+        let Some((final_loc, (final_place, final_rvalue))) = as_assign(result.read, body) else {
+            continue;
+        };
+        // We only apply this optimization within a block, this means that we get to skip all
+        // reasoning about flow control in our analysis.
+        if temp_loc.block != final_loc.block {
+            continue;
+        }
+
+        // If this is a single statement, it's an assignment that modifies a variable based on
+        // itself, such as an AddAssign impl.
+        if temp_loc == final_loc {
+            continue;
+        }
+
+        // If the creation of the temporary comes after the use of it, then by whatever means this
+        // is not the pattern we are looking for.
+        if temp_loc.statement_index > final_loc.statement_index {
+            continue;
+        }
+
+        if invalidated_statements.contains(&temp_loc) || invalidated_statements.contains(&final_loc)
+        {
+            continue;
+        }
+
+        // We only do optimizations where there are no statements other than storage markers
+        // between the two statements. This is our proxy for alias analysis: if there are no
+        // statements between the creation of the temporary and its use, nothing can
+        // observe us eliding the temporary.
+        let statements_are_adjacent = &body.basic_blocks[temp_loc.block].statements
+            [temp_loc.statement_index + 1..final_loc.statement_index]
+            .iter()
+            .all(|s| {
+                matches!(
+                    &s.kind,
+                    StatementKind::Nop
+                        | StatementKind::StorageLive(_)
+                        | StatementKind::StorageDead(_)
+                )
+            });
+        if !statements_are_adjacent {
+            continue;
+        }
+
+        let cx = Context { tcx, local_decls: &body.local_decls, param_env };
+        for opt in &[
+            combine_place_projections,
+            combine_ref_deref,
+            combine_ptr_ptr_cast,
+            combine_ref_addressof,
+        ] {
+            let Some(new_statement) =
+                opt(&cx, temp_place, temp_rvalue, final_place, final_rvalue)
+            else { continue; };
+
+            invalidated_statements.insert(temp_loc);
+            invalidated_statements.insert(final_loc);
+
+            let statements = &mut body.basic_blocks_mut()[temp_loc.block].statements;
```

Comment:
You don't change CFG, so `body.basic_block.as_mut_preserve_cfg()` should be enough.